File: src/main/java/com/proffl/typing/service/WordService.java
package com.proffl.typing.service;

import com.proffl.typing.entity.WordAnalysis;
import com.proffl.typing.entity.WordEntity;

import java.util.List;

public interface WordService {
    List<WordEntity> getAllWords();

    /**
     * 取wordCount个字，
     * @param wordCount 取的数量
     * @param isRepeat 是否允许重复
     * @return 取的字组成的集合
     */
    List<WordEntity> getWords(int wordCount, boolean isRepeat);

    /**
     * 取错误的汉字，如果错误的汉字不足wordCount，复制N份，直到达到wordCount
     * @param wordCount 取的数量,
     * @return
     */
    List<WordEntity> getWrongWords(int wordCount);

    /**
     * 取分析后的word数据
     * @param pageSize
     * @param pageIndex
     * @param filterLetter
     * @return
     */
    List<WordAnalysis> getAnalyzedWords(int pageSize, int pageIndex, String filterLetter);
}

File: src/main/ui/src/app/service/word.service.ts
import {Injectable} from '@angular/core';

import {AppSettings} from '../appsetings';
import {map} from 'rxjs/operators';
import {HttpClient, HttpParams} from '@angular/common/http';
import {WordEntity} from '../shared/entity/word.entity';
import {WordDetailEntity} from '../shared/entity/word-detail.entity';
import {Observable} from 'rxjs/internal/Observable';
import {AnalyzedWordEntity} from '../shared/entity/analyzed-word.entity';

@Injectable({
  providedIn: 'root'
})
export class WordService {
  apiRoot = AppSettings.API_ENDPOINT + 'word';

  constructor(private httpClient: HttpClient) {
  }

  getWords(wordCount: number, isRepeat = false) {
    const params = new HttpParams().set('wordCount', wordCount + '').set('isrepeat', isRepeat + '');

    return this.httpClient
      .get<WordEntity[]>(`${this.apiRoot}/get_words?${params.toString()}`, {})
      .pipe(map(data => {
        const wordDetails: WordDetailEntity[] = new Array();
        data.map(d => {
          let wordDetail: WordDetailEntity = new WordDetailEntity();
          wordDetail.word = d;
          wordDetails.push(wordDetail);
        });
        return wordDetails;
      }));
  }

  getAnalyzedWords(pageSize: number, pageIndex: number, filterLetter: string): Observable<AnalyzedWordEntity[]> {
    return this.httpClient.get<AnalyzedWordEntity[]>(`${this.apiRoot}/get_analyzed_words`, {
        params: new HttpParams()
          .set('pageIndex', pageIndex.toString())
          .set('pageSize', pageSize.toString())
          .set('filterLetter', filterLetter)
    }).pipe(
      map(res => res)
    );
  }
}
